src/com/amazon/ion/impl/IonReaderContinuableApplicationBinary.java

```diff
@@ -0,0 +1,689 @@
+
+        private void finishReadingImportsList() {
+            stepOutOfContainer();
+            imports = new LocalSymbolTableImports(newImports);
+            firstLocalSymbolId = imports.getMaxId() + 1;
+            state = State.ON_SYMBOL_TABLE_FIELD;
+        }
+
+        private void startReadingSymbolsList() {
+            newSymbols = new ArrayList<>(8);
+            state = State.READING_SYMBOL_TABLE_SYMBOLS_LIST;
+        }
+
+        private void startReadingSymbol() {
+            if (IonReaderContinuableApplicationBinary.super.getType() == IonType.STRING) {
+                state = State.READING_SYMBOL_TABLE_SYMBOL;
+            } else {
+                newSymbols.add(null);
+            }
+        }
+
+        private void finishReadingSymbol() {
+            newSymbols.add(stringValue());
+            state = State.READING_SYMBOL_TABLE_SYMBOLS_LIST;
+        }
+
+        private void finishReadingSymbolsList() {
+            stepOutOfContainer();
+            state = State.ON_SYMBOL_TABLE_FIELD;
+        }
+
+        private void startReadingImportStruct() {
+            name = null;
+            version = -1;
+            maxId = -1;
+            if (IonReaderContinuableApplicationBinary.super.getType() == IonType.STRUCT) {
+                stepIntoContainer();
+                state = State.READING_SYMBOL_TABLE_IMPORT_STRUCT;
+            }
+        }
+
+        private void finishReadingImportStruct() {
+            stepOutOfContainer();
+            newImports.add(createImport(name, version, maxId));
+            state = State.READING_SYMBOL_TABLE_IMPORTS_LIST;
+        }
+
+        private void startReadingImportStructField() {
+            int fieldId = getFieldId();
+            if (fieldId == NAME_SID) {
+                state = State.READING_SYMBOL_TABLE_IMPORT_NAME;
+            } else if (fieldId == VERSION_SID) {
+                state = State.READING_SYMBOL_TABLE_IMPORT_VERSION;
+            } else if (fieldId == MAX_ID_SID) {
+                state = State.READING_SYMBOL_TABLE_IMPORT_MAX_ID;
+            }
+        }
+
+        private void readImportName() {
+            if (IonReaderContinuableApplicationBinary.super.getType() == IonType.STRING) {
+                name = stringValue();
+            }
+            state = State.READING_SYMBOL_TABLE_IMPORT_STRUCT;
+        }
+
+        private void readImportVersion() {
+            if (IonReaderContinuableApplicationBinary.super.getType() == IonType.INT) {
+                version = intValue();
+            }
+            state = State.READING_SYMBOL_TABLE_IMPORT_STRUCT;
+        }
+
+        private void readImportMaxId() {
+            if (IonReaderContinuableApplicationBinary.super.getType() == IonType.INT) {
+                maxId = intValue();
+            }
+            state = State.READING_SYMBOL_TABLE_IMPORT_STRUCT;
+        }
+
+        void readSymbolTable() {
+            Event event;
+            while (true) {
+                switch (state) {
```

Comment:
In scanning this I notice there are roughly 3? patterns.

There's the "on item in container or maybe end"?:
```
                        event = IonReaderContinuableApplicationBinary.super.nextValue();
                        if (Event.NEEDS_DATA == event) {
                            return;
                        }
                        if (event == Event.END_CONTAINER) {
                            finishContainerForDealie();
                            return;
                        }
                        readSpecificThingamajig();
                        break;
```

and "read scalar"?:
```
                        if (valueUnavailable()) {
                            return;
                        }
                        readDealieMajig();
                        break;
```

There is some variance and the others are roughly similar but not the same. I wonder if and how we could factor this to avoid repetition. Possibly not without sacrificing performance, I don't know.

Reply:
> In scanning this I notice there are roughly 3? patterns.
> 
> There's the "on item in container or maybe end"?:
> 
> ```
>                         event = IonReaderContinuableApplicationBinary.super.nextValue();
>                         if (Event.NEEDS_DATA == event) {
>                             return;
>                         }
>                         if (event == Event.END_CONTAINER) {
>                             finishContainerForDealie();
>                             return;
>                         }
>                         readSpecificThingamajig();
>                         break;
> ```
> 
> and "read scalar"?:
> 
> ```
>                         if (valueUnavailable()) {
>                             return;
>                         }
>                         readDealieMajig();
>                         break;
> ```
> 
> There is some variance and the others are roughly similar but not the same. I wonder if and how we could factor this to avoid repetition. Possibly not without sacrificing performance, I don't know.

I wrote this when we were still targeting Java 1.6, but yeah, lambdas help us clean this up a bit. Comparing lines of code before and after it's a bit of a wash since we still have to branch for return vs break at each call site, but it probably looks a bit cleaner. I'll include the change in a separate PR.

Reply:
Update, I tried simplifying this ([gist](https://gist.github.com/tgregg/e6fe7a3e1fde3036981bafb830024969)), but found that symbol table parsing performance regressed. Although this will not be noticeable when reading large streams with infrequent symbol table changes, it was noticeable in our benchmark over real-world data that includes a single small value preceded by a symbol table.

I will leave this as-is for now, but feel free to comment on the gist if you notice anything that can be optimized.